clients/src/main/java/org/apache/kafka/common/requests/TxnOffsetCommitRequest.java

```diff
@@ -0,0 +1,188 @@
+
+    @Override
+    protected Struct toStruct() {
+        Struct struct = new Struct(ApiKeys.TXN_OFFSET_COMMIT.requestSchema(version()));
+        struct.set(CONSUMER_GROUP_ID_KEY_NAME, consumerGroupId);
+        struct.set(PID_KEY_NAME, pid);
+        struct.set(EPOCH_KEY_NAME, epoch);
+        struct.set(RETENTION_TIME_KEY_NAME, retentionTimeMs);
+
+        Map<String, Map<Integer, CommittedOffset>> mappedPartitionOffsets = CollectionUtils.groupDataByTopic(offsets);
+        Object[] partitionsArray = new Object[mappedPartitionOffsets.size()];
+        int i = 0;
+        for (Map.Entry<String, Map<Integer, CommittedOffset>> topicAndPartitions : mappedPartitionOffsets.entrySet()) {
+            Struct topicPartitionsStruct = struct.instance(TOPIC_PARTITIONS_KEY_NAME);
+            topicPartitionsStruct.set(TOPIC_KEY_NAME, topicAndPartitions.getKey());
+
+            Map<Integer, CommittedOffset> partitionOffsets = topicAndPartitions.getValue();
+            Object[] partitionOffsetsArray = new Object[partitionOffsets.size()];
+            int j = 0;
+            for (Map.Entry<Integer, CommittedOffset> partitionOffset: partitionOffsets.entrySet()) {
+                Struct partitionOffsetStruct = topicPartitionsStruct.instance(PARTITIONS_KEY_NAME);
+                partitionOffsetStruct.set(PARTITION_KEY_NAME, partitionOffset.getKey());
+                CommittedOffset committedOffset = partitionOffset.getValue();
+                partitionOffsetStruct.set(OFFSET_KEY_NAME, committedOffset.offset);
+                partitionOffsetStruct.set(METADATA_KEY_NAME, committedOffset.metadata);
+                partitionOffsetsArray[j++] = partitionOffsetStruct;
+            }
+            topicPartitionsStruct.set(PARTITIONS_KEY_NAME, partitionOffsetsArray);
+            partitionsArray[i++] = topicPartitionsStruct;
+        }
+
+        struct.set(TOPIC_PARTITIONS_KEY_NAME, partitionsArray);
+        return struct;
+    }
+
+    @Override
+    public TxnOffsetCommitResponse getErrorResponse(Throwable e) {
+        Errors error = Errors.forException(e);
+        Map<TopicPartition, Errors> errors = new HashMap<>(offsets.size());
+        for (TopicPartition partition : offsets.keySet())
+            errors.put(partition, error);
+        return new TxnOffsetCommitResponse(errors);
+    }
+
+    public static TxnOffsetCommitRequest parse(ByteBuffer buffer, short version) {
+        return new TxnOffsetCommitRequest(ApiKeys.TXN_OFFSET_COMMIT.parseRequest(version, buffer), version);
+    }
+
+    public static class CommittedOffset {
+        private final long offset;
+        private final String metadata;
+
+        public CommittedOffset(long offset, String metadata) {
+            this.offset = offset;
+            this.metadata = metadata;
+        }
+
+        @Override
+        public String toString() {
+            return "CommittedOffset(" +
+                    "offset=" + offset +
+                    ", metadata='" + metadata + '\'' +
```

Comment:
Nit: it's a bit annoying to have the `\` escape. It's not required if we use double quotes.